### DISCUSSION TITLE
Non-email authentication

### DISCUSSION THREAD

**Discussion by bittermandel** [fastapi-users/fastapi-users#715] [Ideas]
We are using fastapi-users extensively for https://badgerleague.com - a PoE community event.

In our flow, each user authenticates with their Path of Exile account and later has to add Discord authentication on top of this. 

Unfortunately, Path of Exile does not have an email address scope for their API. We've had to essentially fake the email returned by the oauth_client (PoE username + "@tmpdomain.com") to be able to integrate the rest of the OAuth flow.
This also means that we've had to fork the repository to be able to add the Discord account to the current authenticated user, rather than an account that matches the email returned by `oauth_client.get_id_email`. The fork can be found here: https://github.com/bittermandel/fastapi-users

Is there any plans to remove email as the identifier for users, for us that can't rely on it?

Thank you for this lovely piece of software, we're super happy overall!

PS: There's a similar issue with using Database as a backend, which meant that we had to implement/fake that as well. We're already working on a SQLAlchemy 1.4+ Async backend that we can donate when it's ready, so it's less of an issue :+1:

**Comment by frankie567** [comment]
Unfortunately, I've designed the library with e-mail in mind, so it would be very difficult for me to make it support both approaches. 

However, instead of forking, I would suggest that you implement your own router for your OAuth logic that lives in your project. You can copy/paste the existing one and make the changes you need ; and include this router inside your app. This way, you can keep the common logic of FastAPI Users without worrying about updating the fork and only maintain the part you want to customize.

As for SQLAlchemy 1.4+, I've updated the current backend with a newer version of encode/databases (see https://github.com/fastapi-users/fastapi-users-db-sqlalchemy). I'll make a release for this soon.

**Reply by bittermandel** [reply]
> Unfortunately, I've designed the library with e-mail in mind, so it would be very difficult for me to make it support both approaches.
> 
> However, instead of forking, I would suggest that you implement your own router for your OAuth logic that lives in your project. You can copy/paste the existing one and make the changes you need ; and include this router inside your app. This way, you can keep the common logic of FastAPI Users without worrying about updating the fork and only maintain the part you want to customize.

That's a good idea, we'll do that :)

> As for SQLAlchemy 1.4+, I've updated the current backend with a newer version of encode/databases (see https://github.com/fastapi-users/fastapi-users-db-sqlalchemy). I'll make a release for this soon.

Yes, I saw that you've split them apart into different repos. We don't want to depend on encode/databases at all, so I took the freedom to make a asyncpg one that you can find here: https://github.com/BPL-Development-Team/fastapi-users-db-sqlalchemy-asyncpg

Thanks for the answers :+1:
